### ISSUE TITLE
Possible issue with metrics val_weighted_categorical_accuracy.

### ISSUE THREAD

**Issue by raffaeleabate** [keras-team/keras#7985]
Hi @fchollet . First of all let me thank you for your amazing work on **Keras**. 

I'm using keras=2.0.8 on the tensorflow-gpu=1.3.0 backend.

I'm trying to implement the use of `class_weight` on `model.fit` for a multi-label classification task with metrics `categorical accuracy` and `weighted categorical accuracy`.

I noticed that the output of `val_categorical_accuracy` is always equal to `val_weighted_categorical_accuracy`, which I guess is an incorrect behaviour.

![360](https://user-images.githubusercontent.com/22766463/30817715-e126c718-a219-11e7-84e7-0dd5f1632fb6.png)

My code looks roughly like this:

```
#This model has five classes which have an unbalanced representation of samples. I took the most represented and I rapported the others to it, having value 1
class_weight {0 : 1, 1 : 1.566, 2 : 22.34, 3 : 12.356, 4 : 29.981}


X_full, Y_full = full_set.as_matrix(columns=listColumnsInputs), full_set.as_matrix(columns=listColumnsOutputs)
X_train, Y_train = train_set.as_matrix(columns=listColumnsInputs), train_set.as_matrix(columns=listColumnsOutputs)
X_test, Y_test = test_set.as_matrix(columns=listColumnsInputs), test_set.as_matrix(columns=listColumnsOutputs)

X_full = X_full.reshape(len(X_full), len(listColumnsInputs))
X_train = X_train.reshape(len(X_train), len(listColumnsInputs))
X_test = X_test.reshape(len(X_test), len(listColumnsInputs))

batch_size = 128
nb_classes = len(listColumnsOutputs)

model = Sequential()
num_h = 100
numerolayers = 3
loss_function =  'mean_squared_logarithmic_error'
optimizer_function = 'SGD'

num_epoch = 1000
regularizerl2 = .0
regularizerl1 = .0

model.add(Dense(num_h, input_dim=len(listColumnsInputs), kernel_initializer='uniform', activation='relu', kernel_regularizer=regularizers.l2(regularizerl2), activity_regularizer=regularizers.l1(regularizerl1)))
for x in range(1, numerolayers): model.add(Dense(num_h, kernel_initializer='uniform', activation='relu', kernel_regularizer=regularizers.l2(regularizerl2), activity_regularizer=regularizers.l1(regularizerl1)))

model.add(Dense(len(listColumnsOutputs), kernel_initializer='uniform', activation='softmax'))

model.summary()

model.compile(loss=loss_function,  optimizer=optimizer_function, weighted_metrics=['categorical_accuracy'],  metrics=['categorical_accuracy'])

history = model.fit(X_train, Y_train, class_weight=class_weight, batch_size=batch_size, epochs=num_epoch, verbose=1, validation_data=(X_test, Y_test), callbacks=[TensorBoard(log_dir='./logs', histogram_freq=25, write_graph=True, write_images=False)])

score = model.evaluate(X_test, Y_test, verbose=0)
```

Is this a small bug? Am I wrong? 

Thanks a lot for your time.

**Comment by nicolewhite** [comment]
This is my bug most likely. I'll take a look.

**Comment by raffaeleabate** [comment]
Thank you, Nicole.

**Comment by nicolewhite** [comment]
I can't reproduce. :( I always get different weighted and unweighted metrics. Is it possible your validation data consists only of class 0?

**Comment by raffaeleabate** [comment]
>  I always get different weighted and unweighted metrics. 

In both `weighted_categorical_accuracy` and `val_weighted_categorical_accuracy`? 

> Is it possible your validation data consists only of class 0?

Unfortunately not :( That would explain it but I just double checked.

Currently I have six classes (from **0** to **5**) with these amounts of samples:

**Training samples →** `{0 : 150104, 1 : 5092, 2 : 14056, 3 : 14260, 4 : 16150, 5 : 14518}`

**Validation samples →** `{0 : 30004, 1 : 1018, 2 : 2809, 3 : 2851, 4 : 3228, 5 : 2901}`

**Applied Weights →** `{0: 1.0, 1: 29.48, 2: 10.68, 3: 10.53, 4: 9.29, 5: 10.34}`


I want to point out that the model is able of generalization at this point - also for the  less represented classes. A simple `classification report` made on the validation set looks like this:

             precision    recall  f1-score   support

          N       0.99      1.00      0.99     30004
          A       0.93      0.90      0.91      1018
          /       1.00      1.00      1.00      2809
          V       0.98      0.96      0.97      2851
          L       0.99      1.00      0.99      3228
          R       1.00      1.00      1.00      2901

`avg / total       0.99      0.99      0.99     42811
`

The `weighted_categorical_accuracy` seems to work fine on my side, the only problem I've got is with the `val_weighted_categorical_accuracy` which is stuck as if I didn't use the weighted metrics parameter.

![357](https://user-images.githubusercontent.com/22766463/30956454-ad48917c-a437-11e7-9d40-5bbcd7d2f666.png)

I'm searching in https://github.com/fchollet/keras/blob/master/keras/engine/training.py but I still cannot see what might be wrong.

If there are updates I'll let you know.

Thanks for your help.

**Comment by fuine** [comment]
I can reproduce that bug, but with both `categorical_accuracy` and `weighted_categorical_accuracy` as well as `val_categorical_accuracy` and `val_weighted_categorical_accuracy`. I can cooperate and help debugging if any of the maintainers is willing to take a look.

**Comment by nicolewhite** [comment]
Hey all, I see the issue. You are meant to pass in the validation sample weights as the third part of the `validation_data` tuple. See [the fit docs](https://keras.io/models/sequential/#fit):

> validation_data: tuple (x_val, y_val) or tuple (x_val, y_val, val_sample_weights) on which to evaluate the loss and any model metrics at the end of each epoch.

A full working example of this:

```python
import numpy as np

from keras import regularizers
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

num_train_samples = 1000
num_test_samples = 100
num_classes = 5
num_feats = 10
batch_size = 12
num_hidden = 100
num_layers = 3

x_train = np.random.normal(size=(num_train_samples, num_feats))
x_test = np.random.normal(size=(num_test_samples, num_feats))

y_train = np.random.randint(num_classes, size=(num_train_samples))
y_test = np.random.randint(num_classes, size=(num_test_samples))

print('x train shape', x_train.shape)
print('x test shape', x_test.shape)
print('y train shape', y_train.shape)
print('y test shape', y_test.shape)
```

```
x train shape (1000, 10)
x test shape (100, 10)
y train shape (1000,)
y test shape (100,)
```

```python
model = Sequential()

for i in range(num_layers):
    kwargs = {
        'units': num_hidden,
        'kernel_initializer': 'uniform',
        'activation': 'relu',
        'kernel_regularizer': regularizers.l2(0.),
        'activity_regularizer': regularizers.l1(0.)
    }
    if i == 0:
        kwargs['input_shape'] = (num_feats,)
        
    model.add(Dense(**kwargs))

model.add(Dense(units=num_classes, kernel_initializer='uniform', activation='softmax'))

model.summary()
```

```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense_1 (Dense)              (None, 100)               1100      
_________________________________________________________________
dense_2 (Dense)              (None, 100)               10100     
_________________________________________________________________
dense_3 (Dense)              (None, 100)               10100     
_________________________________________________________________
dense_4 (Dense)              (None, 5)                 505       
=================================================================
Total params: 21,805
Trainable params: 21,805
Non-trainable params: 0
_________________________________________________________________
```

```python
model.compile(
    loss='mean_squared_logarithmic_error',  
    optimizer='SGD', 
    weighted_metrics=['categorical_accuracy'],  
    metrics=['categorical_accuracy']
)

class_weight = {0 : 1, 1 : 1.566, 2 : 22.34, 3 : 12.356, 4 : 29.981}
val_sample_weight = np.array([class_weight[cls] for cls in y_test])

y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

print('val sample weight shape', val_sample_weight.shape)
print('y train shape', y_train.shape)
print('y test shape', y_test.shape)
```

```
val sample weight shape (100,)
y train shape (1000, 5)
y test shape (100, 5)
```

```python
model.fit(
    x=x_train, 
    y=y_train, 
    class_weight=class_weight, 
    batch_size=12, 
    epochs=5, 
    verbose=2, 
    validation_data=(x_test, y_test, val_sample_weight)
)
```

```
Train on 1000 samples, validate on 100 samples
Epoch 1/5
 - 1s - loss: 0.9055 - categorical_accuracy: 0.2080 - weighted_categorical_accuracy: 6.2360 - val_loss: 0.9016 - val_categorical_accuracy: 0.1800 - val_weighted_categorical_accuracy: 5.3966
Epoch 2/5
 - 0s - loss: 0.8955 - categorical_accuracy: 0.2080 - weighted_categorical_accuracy: 6.2360 - val_loss: 0.8968 - val_categorical_accuracy: 0.1800 - val_weighted_categorical_accuracy: 5.3966
Epoch 3/5
 - 0s - loss: 0.8882 - categorical_accuracy: 0.2080 - weighted_categorical_accuracy: 6.2360 - val_loss: 0.8935 - val_categorical_accuracy: 0.1800 - val_weighted_categorical_accuracy: 5.3966
Epoch 4/5
 - 0s - loss: 0.8828 - categorical_accuracy: 0.2080 - weighted_categorical_accuracy: 6.2360 - val_loss: 0.8912 - val_categorical_accuracy: 0.1800 - val_weighted_categorical_accuracy: 5.3966
Epoch 5/5
 - 0s - loss: 0.8786 - categorical_accuracy: 0.2080 - weighted_categorical_accuracy: 6.2360 - val_loss: 0.8895 - val_categorical_accuracy: 0.1800 - val_weighted_categorical_accuracy: 5.3966
```

**Comment by keifler** [comment]
this addresses how to use 'sample_weight', but it doesn't fix 'class_weight' during validation.
